examples/llama.py

```diff
@@ -181,27 +170,86 @@ def sample(logits, temperature):
 def concat_weights(models):
   def convert(name) -> Tensor:
     disk_tensors = [model[name] for model in models]
-    if len(disk_tensors) == 1:
-      return disk_tensors[0]
-    if len(disk_tensors[0].shape) == 1:
-      return disk_tensors[0]
-    if name.startswith('tok_embeddings.') or name.endswith('.attention.wo.weight') or name.endswith('.feed_forward.w2.weight'):
-      axis = 1
-    else:
-      axis = 0
+    if len(disk_tensors) == 1 or len(disk_tensors[0].shape) == 1:
+      return disk_tensors[0].to(device=Device.DEFAULT)
+    axis = 1 if name.startswith('tok_embeddings.') or name.endswith('.attention.wo.weight') or name.endswith('.feed_forward.w2.weight') else 0
     lazy_tensors = [data.to(device=Device.DEFAULT) for data in disk_tensors]
-    first, rest = lazy_tensors[0], lazy_tensors[1:]
-    return first.cat(*rest, dim=axis)
+    return lazy_tensors[0].cat(*lazy_tensors[1:], dim=axis)
   return {name: convert(name) for name in {name: None for model in models for name in model}}
 
+class AbsmaxQuantizedLinear:
+  def __init__(self, in_features, out_features):
+    self.weight = Tensor.ones(out_features, in_features, dtype=dtypes.int8)
+    self.scale = Tensor.ones(out_features, dtype=dtypes.half)
+
+  def __call__(self, x):
+    return x.dot(self.weight.cast(dtype=dtypes.half).T/self.scale)
+
+  @staticmethod
+  def quantize(tensors):
+    new_tensors = {}
+    for name,v in tensors.items():
+      if 'feed_forward' in name or ('attention.w') in name or name == 'output.weight':
+        scale = 127.0 / v.abs().max(axis=1)
+        int8_weight = (v.T*scale).T.cast(dtype=dtypes.int8)
+        new_tensors[name] = int8_weight
+        new_tensors[name.replace('weight', 'scale')] = scale
+      else:
+        new_tensors[name] = v
+    return new_tensors
+
+class LLaMa:
+  @staticmethod
+  def build(model_path, tokenizer_path, model_size="7B", quantize=False):
+    from sentencepiece import SentencePieceProcessor
+    sp_model = SentencePieceProcessor(model_file=str(tokenizer_path))
+    assert sp_model.vocab_size() == VOCAB_SIZE
+
+    from tinygrad.state import torch_load, load_state_dict
+    if model_size == "65B":
+      model = Transformer(**args_65B, linear=AbsmaxQuantizedLinear) if quantize else Transformer(**args_65B)
+      num_files = 8
+    elif model_size == "30B":
+      model = Transformer(**args_30B, linear=AbsmaxQuantizedLinear) if quantize else Transformer(**args_30B)
+      num_files = 4
+    elif model_size == "13B":
+      model = Transformer(**args_13B, linear=AbsmaxQuantizedLinear) if quantize else Transformer(**args_13B)
+      num_files = 2
+    else:
+      model = Transformer(**args_7B, linear=AbsmaxQuantizedLinear) if quantize else Transformer(**args_7B)
+      num_files = 1
+
+    weights = concat_weights([torch_load(filename) for filename in [f"{model_path}/{model_size}/consolidated.{i:02d}.pth" for i in range(num_files)]])
```

Comment:
Want to confirm this is streaming...aka it quantizes the layers as it loads them. I want to avoid high peak memory usage.

Reply:
Confirmed.  Memory increases linearly and peaks at the end once all the weights are loaded.